offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy

```diff
@@ -0,0 +1,370 @@
+     * <p>Please note: It is important that the template document provides the following css selectors</p>
+     * <ul>
+     *  <li>.currency-symbol</li>
+     *  <li>.vat-percentage</li>
+     *  <li>.overheads-percentage</li>
+     *  <li>#customer-name</li>
+     *  <li>#customer-organisation</li>
+     *  <li>#customer-street</li>
+     *  <li>#customer-postal-code</li>
+     *  <li>#customer-city</li>
+     *  <li>#customer-country</li>
+     *  <li>#qbic-ad</li>
+     *  <li>#project-manager-name</li>
+     *  <li>#project-manager-street</li>
+     *  <li>#project-manager-city</li>
+     *  <li>#project-manager-email</li>
+     *  <li>#doctype-desc</li>
+     *  <li>#offer-date</li>
+     *  <li>#offer-identifier</li>
+     *  <li>#offer-expiry-date</li>
+     *  <li>#project-title</li>
+     *  <li>#project-description</li>
+     *  <li>#experimental-design</li>
+     *  <li>.overheads  .data-generation  > .costs</li>
+     *  <li>.overheads  .data-analysis  > .costs</li>
+     *  <li>.overheads  .data-management > .costs</li>
+     *  <li>.overheads  > .totals-section  > .costs</li>
+     *  <li>.total-costs .net > .costs</li>
+     *  <li>.total-costs .discounts > .costs</li>
+     *  <li>.total-costs .overheads > .costs</li>
+     *  <li>.total-costs .vat > .costs</li>
+     *  <li>.total-costs > .totals-section > .costs</li>
+     *  <li>#data-generation-items</li>
+     *  <li>#data-generation-items > tfoot .costs</li>
+     *  <li>#data-analysis-items</li>
+     *  <li>#data-analysis-items > tfoot .costs</li>
+     *  <li>#data-management-items</li>
+     *  <li>#data-management-items > tfoot .costs</li>
+     * </ul>
+     * @param document the template document
+     * @since 1.0.0
+     */
+    OfferTemplate(Document document) {
+        this.document = document
+    }
+
+    /**
+     * Fills the template with the specified content
+     * @param content the content to be parsed and replaced in the offer
+     * @return a document filled with the provided content based on this template
+     * @since 1.1.0
+     */
+    Document fill(OfferContent content) {
+        // make a copy of the template
+        Document result = this.document.clone()
+        // fill customer information
+        fillCustomerInformation(result, content)
+        // fill project manager information
+        fillProjectManagerInformation(result, content)
+        // fill offer information
+        fillOfferInformation(result, content)
+        // fill cost summaries
+        fillCostSummary(result, content)
+        // fill tables table
+        fillDataGenerationItems(result, content)
+        fillDataAnalysisItems(result, content)
+        fillDataManagementItems(result, content)
+        // fill general information
+        fillGeneralInformation(result, content)
+        return result
+    }
+
+    private static void fillGeneralInformation(Document document, OfferContent offer) {
```

Comment:
> why is `document` provided through the method? Why not use the class variable you introduced? And make the method non-static
> 
> the same for all the other methods

This is because we generate a new document from the template and don't modify the template itself.

Reply:
> Maybe add a hint so that is clear that this method should be run at the end after all information is filled? Because you add the currency then. But I also dont like that it's done like this. Why not handling that with a Formatter and forward the price already with the currency?

Because that adds complexity we do not need. And because we hard-coded the currency symbol in the HTML and css before. This is now no longer the case and is handled by the business object Currency.

Reply:
> fill general might not really describes what you do here, its more like fillSpecialCharacter.

We do not only set special characters here. What we do is, we set offer global variables like overhead percentage, vat percentage and currency symbol

Reply:
But if you introduce a class variables for the Document and OfferContent (and overwrite them each time fill() is called) you would improve readability of the code and make the methods easier to read. I think the parameters are not necessary here. Its a CleanCode thing to try to reduce the parameters and improve the readability

Reply:
It is very important to reduce side effects and the readability is not decreased by this I think.